example_allocators/compartment_alloc/compartment_alloc.c

```diff
@@ -0,0 +1,91 @@
+#include <cheriintrin.h>
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/mman.h>
+#include "compartment_alloc.h"
+#include "utils.h"
+
+heap_compartment compartments[maxCompartments];
+int numCompartments = 0;
+
+/*
+ * Initialise a compartment with `size_in_bytes` space, where all future
+ * allocations will be constrained by `dc` as if it was the DDC when allocation
+ * occurred.
+ */
+void
+*__capability init_compartment(int size_in_bytes, void *__capability dc)
+{
+    // We impose a limit on our # of compartments
+    if (numCompartments + 1 == maxCompartments) {
+        return NULL;
+    }
+
+    // Allocate memory for this compartment
+    void *buf = mmap(NULL, size_in_bytes, PROT_READ | PROT_WRITE, MAP_ANON | MAP_PRIVATE, -1, 0);
+    if (buf == MAP_FAILED)
+    {
+        perror("error in initial mem allocation");
+        exit(-1);
+    }
+
+    // Create the new compartment
+    compartments[numCompartments] = (heap_compartment){
+        buf,                     // Space on the heap
+        0,                       // bytes allocated (currently none!)
+        size_in_bytes,           // Our maximum size, to check when allocating
+        dc,                      // The data capability to allocate against (like a DDC)
+        sealed_reference(buf)    // An identifier which allows a holder to allocate in this component
```

Comment:
Where is `sealed_reference` coming from?

Reply:
Ah rats! It's coming from `utils.h`, which I completely missed when adding files to the PR. Sorry…!

Fixed in [0ab49e8](https://github.com/capablevms/cheri-examples/pull/88/commits/0ab49e82383482093c8ba8857525f4777daba621).